radix-engine/src/model/substates/substate.rs

```diff
@@ -266,3 +261,337 @@ impl Into<GlobalAddressSubstate> for Substate {
+        node_pointer: RENodePointer,
+        offset: SubstateOffset,
+        prev_children: HashSet<RENodeId>,
+        call_frames: &'f mut Vec<CallFrame>,
+        track: &'f mut Track<'s, R>,
+    ) -> Result<Self, RuntimeError> {
+        let substate_ref_mut = Self {
+            lock_handle,
+            prev_children,
+            node_pointer,
+            offset,
+            call_frames,
+            track,
+        };
+        Ok(substate_ref_mut)
+    }
+
+    pub fn offset(&self) -> &SubstateOffset {
+        &self.offset
+    }
+
+    pub fn flush(&mut self) -> Result<(), RuntimeError> {
+        let (new_global_references, new_children) = {
+            let substate_ref_mut = self.get_raw_mut();
+            substate_ref_mut.to_ref().references_and_owned_nodes()
+        };
+
+        let current_frame = self.call_frames.last_mut().unwrap();
+
+        for global_address in new_global_references {
+            let node_id = RENodeId::Global(global_address);
+            if !current_frame.node_refs.contains_key(&node_id) {
+                return Err(RuntimeError::KernelError(
+                    KernelError::InvalidReferenceWrite(global_address),
+                ));
+            }
+        }
+
+        // Take values from current frame
+        let (taken_nodes, missing_nodes) = {
+            if !new_children.is_empty() {
+                if !SubstateProperties::can_own_nodes(&self.offset) {
+                    return Err(RuntimeError::KernelError(KernelError::ValueNotAllowed));
+                }
+
+                current_frame.take_available_values(new_children, true)?
+            } else {
+                (HashMap::new(), HashSet::new())
+            }
+        };
+        verify_stored_value_update(&self.prev_children, &missing_nodes)?;
+
+        for child_node in taken_nodes.keys() {
+            current_frame.add_lock_visible_node(self.lock_handle, *child_node)?;
+        }
+        self.node_pointer
+            .add_children(taken_nodes, &mut self.call_frames, &mut self.track);
+
+        Ok(())
+    }
+
+    pub fn overwrite(&mut self, substate: Substate) -> Result<(), RuntimeError> {
+        let raw_mut = self.get_raw_mut();
+        match (raw_mut, substate) {
+            (RawSubstateRefMut::ComponentState(current), Substate::ComponentState(next)) => {
+                *current = next
+            }
+            (
+                RawSubstateRefMut::KeyValueStoreEntry(current),
+                Substate::KeyValueStoreEntry(next),
+            ) => *current = next,
+            (RawSubstateRefMut::NonFungible(current), Substate::NonFungible(next)) => {
+                *current = next
+            }
+            (RawSubstateRefMut::System(current), Substate::System(next)) => *current = next,
+            _ => return Err(RuntimeError::KernelError(KernelError::InvalidOverwrite)),
+        }
+
+        self.flush()
+    }
+
+    pub fn get_raw_mut(&mut self) -> RawSubstateRefMut {
```

Comment:
Why do we need two write interfaces, `overwrite` and `get_raw_mut`? With a raw mutable reference, we can already overwrite the value.

Reply:
`overwrite` was really a convenience method to "flash" a substate. Will see if it makes sense to remove this.